components/list/Lists.tsx

```diff
@@ -21,6 +29,37 @@ export function Lists({ actor }: { actor: string }) {
     getNextPageParam: (lastPage) => lastPage.cursor,
   })
 
+  const lists = data?.pages.flatMap((page) => page.lists)
+
+  const confirmAddToWorkspace = async () => {
+    // add items that are already loaded
+    if (!lists?.length) return
+    await addItemsToWorkspace(lists.map((f) => f.uri))
+    if (!data?.pageParams) {
+      setIsConfirmationOpen(false)
+      return
+    }
+    setIsAdding(true)
+
+    try {
+      let cursor = data.pageParams[0] as string | undefined
+      do {
+        const nextLists = await labelerAgent.app.bsky.graph.getLists({
+          actor,
+          limit: 25,
+          cursor,
+        })
+        await addItemsToWorkspace(nextLists.data.lists.map((f) => f.uri))
+        cursor = nextLists.data.cursor
+        //   if the modal is closed, that means the user decided not to add any more user to workspace
+      } while (cursor && isConfirmationOpen)
```

Comment:
Due to the nature of how react works, this condition will not actually cancel the loop as `isConfirmationOpen` is a const that will never become false.

For this to work you'll probably need to rely on a ref & AbortController.


```ts
const abortRef = ref<AbortController | null>(null)
const isWorking = abortRef.current != null

const startWorking = async () => {
  // Abort previous run, if any
  abortRef.current?.abort()

  const controller = new AbortController()
  abortRef.current = controller

  try {
    do {
      await someAsyncStuff({
        signal: controller.signal
      })
      
      if (controller.signal.aborted) {
        // make sure the catch block is triggered
        throw controller.signal.reason
      }
    } while (true)
  } catch (err) {
    if (controller.signal.aborted) {
      toast.info('this action was aborted')
    } else {
      toast.error(`Something went wrong: ${(e as Error).message}`)
    }
  } finally {
    // Prevent re-use (for good measure)
    controller.abort()

    // If this run is still the currently active one
    if (abortRef.current === controller) {
      // Cleanup reference
      abortRef.current = null

      // Close the modal
      setIsConfirmationOpen(false)
    }
  } 
}

const abortCurrentRun = () => {
  // No need to call setIsConfirmationOpen here as this will be done through the async process
  abortRef.current?.abort()
}

// Abort the current run when the component is unmounted (e.g. due to navigation)
useEffect(() => {
  return () => abortRef.current?.abort()
}, [abortRef])

```

PS: If this feels too manual, the mutation lib probably already does all of the above

Reply:
yeah I knew it wasn't optimal but it does kill the requests if other conditions cause a re-render.
we use this pattern in a few places so I cleaned up all usage of this pattern to use abort controllers. thank you for pointing it out!